MPI support with splitting based on spectra IDs

SimonHeybrock opened the pull request:
This pull request is meant for discussing design and implementation of the new MPI model.

SimonHeybrock committed:
Re #14864 Basic helpers functions for upcoming MPI stuff.

- MPI communicator in FrameworkManager
- function to get rank, root rank, etc.
- enums for StorageMode and Execution mode, to be used in Workspace and
  Algorithm

SimonHeybrock committed:
Re #14864 Tests for MPI helpers functions (only for single-rank run).

These tests do not work/pass when run with more than a single MPI rank.

SimonHeybrock committed:
Re #14864 Workspace carries a MPI::StorageMode flag now.

This will be used by algorithms to figure out the correct
MPI::ExecutionMode.

SimonHeybrock committed:
Re #14864 Basic version of Algorithm "MPI::ExecutionMode" mechanism.

SimonHeybrock committed:
Re #14864 MPI package included also in non-MPI builds.

SimonHeybrock committed:
Re #14864 Fix: ifdef for non-MPI builds with boost::mpi stuff.

SimonHeybrock committed:
Re #14864 Added MPI module to non-Framework builds.

SimonHeybrock committed:
Re #14864 Fix: Disabled test of MPI helper function in non-MPI build.

SimonHeybrock committed:
Re #14864 Improved log and error messages in Algorithm MPI code.

SimonHeybrock committed:
Re #14864 Fix: Checking for nullptr in Algorithm MPI code.

SimonHeybrock committed:
Re #14864 Fix: Checking for more nullptrs in Algorithm MPI code.

SimonHeybrock committed:
Re #14864 MPI::Helpers test ifdef fix for non-MPI build.

SimonHeybrock committed:
Re #14864 Moved MPI directory to new code directory.

SimonHeybrock committed:
Re #14864 Add TriviallyParallelAlgorithm (new Algorithm base class).

SimonHeybrock commented:
I now pushed also `TriviallyParallelAlgorithm`. This class would facilitate the MPI support for 90% of our algorithms. The only required change would be to inherit from `TriviallyParallelAlgorithm` instead of `Algorithm`.

SimonHeybrock committed:
Re #14864 Disable failing unit validation on non-master ranks.

Since the master rank still valides this should be ok.

OwenArnold commented on Framework/API/inc/MantidAPI/Algorithm.h:
```diff
@@ -306,6 +307,34 @@ class MANTID_API_DLL Algorithm : public IAlgorithm,
 
+  virtual void execDistributed() { exec(); }
+  virtual void execMasterOnly();
+  virtual void execNonMaster(){};
```
I'm sure there's a logical explanation for this: How is non-master different from distributed?

SimonHeybrock replied:
Non-master refers to all ranks *but* the master rank. This function is executed on non-master ranks, when to master calls `exec()`. I needed this in 1 or 2 special cases, where despite having no data (since all data is on the master rank), the other ranks still have to do something (like initialize and output workspace).

SimonHeybrock replied:
To clarify this a bit more: If a workspace has `StorageMode::MasterOnly`, the workspace still exists on *all* ranks, but there are not spectra on the non-master ranks. The reason for this is that then we can reuse existing Python reduction scripts without modification. It the workspace would really exist only on the master rank, each Python script would need to check for the rank everywhere and call algorithms only when the workspace is available.

OwenArnold commented on Framework/API/inc/MantidAPI/Algorithm.h:
```diff
@@ -306,6 +307,34 @@ class MANTID_API_DLL Algorithm : public IAlgorithm,
+   * with multiple MPI ranks. May not return ExecutionMode::Serial, because that
+   * is not a "parallel" execution mode. */
+  virtual MPI::ExecutionMode getParallelExecutionMode(
+      const std::map<std::string, MPI::StorageMode> &storageModes) const;
```
What is the key of type std::string representing in the map?

SimonHeybrock replied:
See comment on this related function:
```cxx
   /* The key to the name is the property name of the respective workspace. */
  std::map<std::string, MPI::StorageMode> getInputWorkspaceStorageModes() const;
```

OwenArnold commented on Framework/API/inc/MantidAPI/Algorithm.h:
```diff
@@ -306,6 +307,34 @@ class MANTID_API_DLL Algorithm : public IAlgorithm,
+   * is identified by the name if its property in the Algorithm (*not* the name
+   * of the Workspace). */
+  virtual MPI::StorageMode
+  getStorageModeForOutputWorkspace(const std::string &propertyName) const;
```
Presumably this handles having multiple workspaces with Direction Output.

SimonHeybrock replied:
Yes, `propagateWorkspaceStorageMode()` will call `getStorageModeForOutputWorkspace` for each output.

OwenArnold commented on Framework/MPI/inc/MantidMPI/Types.h:
```diff
@@ -0,0 +1,42 @@
+ * Cloned: There is a copy (clone) of the Workspace on each rank.
+ * Distributed: Each rank holds parts of the Workspace (spectra).
+ * MasterOnly: The master/root rank has the Workspace */
+enum class StorageMode { Cloned, Distributed, MasterOnly };
```
This is good

OwenArnold commented on Framework/API/inc/MantidAPI/TriviallyParallelAlgorithm.h:
```diff
@@ -0,0 +1,54 @@
+/**
+ Base class for algorithms that treat all spectra independently, i.e., we can
+ trivially parallelize over the spectra without changes. The assumption is that
+ we have one input and one output workspace. The storage mode is just propagated
```
Good

OwenArnold commented on Framework/API/src/Algorithm.cpp:
```diff
@@ -700,6 +706,29 @@ void Algorithm::executeAsChildAlg() {
+  if (MPI::isRoot())
+    return exec();
+  else
+    return execNonMaster();
```
Not entirely sure what this is going to do. This has to do with your previous comment regarding setting properties (We'll talk on the phone about this)

OwenArnold commented on Framework/API/inc/MantidAPI/Algorithm.h:
```diff
@@ -306,6 +307,34 @@ class MANTID_API_DLL Algorithm : public IAlgorithm,
+
+  /** Get correct execution mode based on input storage modes for an MPI run.
+   *
+   * The default implementation returns ExecutionMode::Invalid. Classes
```
So entirely up to client algorithm to set this (unless they are trivialalgorithms?)

SimonHeybrock replied:
Correct. I don't think there is an automatic way that is correct in general.

OwenArnold commented on Framework/API/src/Algorithm.cpp:
```diff
@@ -1579,6 +1608,89 @@ void Algorithm::setAlgStartupLogging(const bool enabled) {
   return m_isAlgStartupLoggingEnabled;
 }
+
+MPI::ExecutionMode Algorithm::getExecutionMode() const {
```
Might be worth sticking some tests over this

OwenArnold commented on Framework/API/src/Algorithm.cpp:
```diff
@@ -1579,6 +1608,89 @@ void Algorithm::setAlgStartupLogging(const bool enabled) {
+}
+
+MPI::ExecutionMode
+Algorithm::getCorrespondingExecutionMode(MPI::StorageMode storageMode) const {
```
Placeholder for me to ask you about this

OwenArnold commented on Framework/MPI/inc/MantidMPI/Helpers.h:
```diff
@@ -0,0 +1,57 @@
+#ifndef MANTID_MPI_HELPERS_H_
```
I'm not a big fan of helpers either functions, headers or classes. Could a more descriptive name be used, or could this be broken up into separate headers if needed.

OwenArnold commented on Framework/MPI/inc/MantidMPI/Types.h:
```diff
@@ -0,0 +1,42 @@
+ *
+ * Invalid: Indicates a state where execution is not possible.
+ * Serial: Serial execution (non-MPI build MPI build with single rank).
+ * Identical: Independent execution in the same way on each rank.
```
Placeholder for us to discuss 'Identical'

SimonHeybrock committed:
Re #14864 Usage examples for algorithms in an MPI scenario.

SimonHeybrock committed:
Re #14864 Algorithm example with actual MPI calls.

SimonHeybrock committed:
Re #14864 Temporary commit for testing.

This commit is a dump of all changes not commited so for, including some
noise like debugging printfs, etc. This is just temporary and needs to
be redone.

SimonHeybrock committed:
Re #14864 Added python scripts for testing new MPI stuff.

SimonHeybrock committed:
Re #14864 Removed unused bits from CMakeLists.txt.

SimonHeybrock committed:
Re #14864 Mini fix: disabled incomplete test code that breaks stuff.

SimonHeybrock commented:
This concept will be revived (properly implemented) here: https://github.com/mantidproject/mantid/issues/19246.
